filebeat/module/iis/w3svc/ingest/pipeline.json

```diff
@@ -0,0 +1,46 @@
+{
+  "description": "Pipeline for parsing iis w3svc logs",
+  "processors" : [
+    {
+      "grok" : {
+        "field" : "message",
+        "patterns" : [
+          "%{TIMESTAMP_ISO8601:iis.access.time} %{IPORHOST:iis.access.site} %{WORD:iis.access.method} %{URIPATH:iis.access.url} %{NOTSPACE:iis.access.querystring} %{NUMBER:iis.access.port} %{NOTSPACE:iis.access.user_name} %{IPORHOST:iis.access.remote_ip} %{NOTSPACE:iis.access.agent} ?%{NOTSPACE:iis.access.referrer}? %{NUMBER:iis.access.response_code} %{NUMBER:iis.access.subresponse_code} %{NUMBER:iis.access.scstatus} %{NUMBER:iis.access.response_time}"
+        ]
+      }
+    }, {
+      "rename" : {
+        "field": "@timestamp",
+        "target_field": "read_timestamp"
+    }, {
+      "date" : {
+        "field" : "iis.access.time",
+        "target_field": "@timestamp",
+        "formats" : [
+          "yyyy-MM-dd HH:mm:ss"
+        ]
+      }
+    }, {
+      "user_agent": {
```

Comment:
IIS replaces spaces in the raw user agent string with `+` characters, since fields in IIS logs are space separated. Do we need to handle this in the `grok` pattern, or a subsequent processor? Perhaps replacing these `+` with spaces again so that we get the correct user agent string?

Note that we'll need to consider this carefully, since some user agent strings might actually contain `+` characters (see, for example, the [Googlebot user agent string](https://support.google.com/webmasters/answer/1061943?hl=en)). I'm not entirely sure what IIS does in this case, but I think it leaves any `+` unchanged. Therefore it might not be possible to set this replacement up correctly in all cases.

Reply:
@cgwrench are the user agent URL encoded?

We could use either of https://www.elastic.co/guide/en/elasticsearch/reference/master/gsub-processor.html
or https://www.elastic.co/guide/en/elasticsearch/reference/master/urldecode-processor.html

Reply:
@ph Thanks for the suggestions. I don't think that the the user agent string is URL encoded. IIS replaces spaces with a `+`, rather than `%20`. Also, IIS doesn't replace `+` characters with `%2B`, it leaves them as is, hence my concern above about not being able to unambiguously "undo" whatever encoding IIS is using here. I did take a look at using the gsub processor to replace `+`s with spaces. However, from a bit of testing I've done since starting this review, it looks like the ingest user-agent plugin handles user agent strings with `+`s instead of spaces just fine. So perhaps we don't need to make any changes here.